src/jit/emitxarch.cpp

```diff
@@ -145,6 +145,31 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
     return ((CodeGenInterface::instInfo[ins] & INS_Flags_IsDstSrcSrcAVXInstruction) != 0) && IsAVXInstruction(ins);
 }
 
+bool emitter::doesZeroExtendingWrite(instrDesc* id, regNumber reg)
+{
+    insFormat fmt = id->idInsFmt();
+
+    switch (fmt)
+    {
+        case IF_RWR_CNS:
+        case IF_RRW_CNS:
+        case IF_RRW_SHF:
+        case IF_RWR_RRD:
+        case IF_RRW_RRD:
+        case IF_RWR_MRD:
+        case IF_RWR_SRD:
+        case IF_RWR_ARD:
+
+            // Can't rely on a "small" movsx as we will over-extend to 8 bytes
+            return (id->idIns() != INS_movsx) && (id->idReg1() == reg) && (id->idOpSize() != EA_8BYTE);
```

Comment:
The size check doesn't seem right - x64 only zero extends 32 bit register writes, `mov al, 42` does not zero extend. So the check should be ` == EA_4BYTE`, though that will probably require special casing for `movzx` where size has a different meaning.

Reply:
Good point.

For `movzx` you mean allow both 4 and 8 byte sizes...?

Reply:
For `movzx` and `movsx` the size indicates the source operand size - 1 or 2 bytes - rather than the destination operand size. So `movzx` always zeroes out the upper 32 bit, no matter the size.

It's a bit unfortunate. I'm considering changing movzx/movsx to movzxb/movzxw/movsxb/movsxw in the future, that might simplify things and perhaps allow more code to be shared with ARM.